{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "import os\n",
    "from Bio import SeqIO as seqio\n",
    "import zipfile\n",
    "import numpy as np\n",
    "from kinaid.session import Session\n",
    "from kinaid.matching import MatchWithMapping, Scoring, PeptideBackground, PWM_Matrices\n",
    "from kinaid.utility import Utility\n",
    "from kinaid.ortholog import OrthologManager\n",
    "\n",
    "def download_file(url : str,\n",
    "                filename : str):\n",
    "    headers = {\n",
    "        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36'\n",
    "    }\n",
    "    with requests.get(url, headers=headers, stream=True) as response:\n",
    "        response.raise_for_status()\n",
    "        with open(filename, 'wb') as file:\n",
    "            for chunk in response.iter_content(chunk_size=8192): \n",
    "                if chunk: \n",
    "                    file.write(chunk)\n",
    "\n",
    "def get_protein_sequence(seq, phospho_position):\n",
    "\n",
    "    final_seq = None\n",
    "    if phospho_position < 5 :\n",
    "        final_seq =  '_'*(5 - phospho_position) + seq[0:phospho_position+6]\n",
    "    elif phospho_position > len(seq) - 6:\n",
    "        final_seq =  seq[phospho_position-5:] + '_'*(phospho_position - len(seq) + 6)  \n",
    "    else:\n",
    "        final_seq = seq[phospho_position-5:phospho_position+6]\n",
    "    \n",
    "    if final_seq[5] !='S' and final_seq[5] != 'T' and final_seq[5] != 'Y':\n",
    "        print(final_seq)\n",
    "        return None\n",
    "    return final_seq\n",
    "\n",
    "data_dir = './data'\n",
    "test_dir = './test'\n",
    "proteomes_dir = os.path.join(test_dir, 'proteomes')\n",
    "\n",
    "if not os.path.exists(test_dir):\n",
    "    os.makedirs(test_dir)\n",
    "    \n",
    "johnson_ST_matrices_file = os.path.join(data_dir,'ST-Kinases.xlsx')\n",
    "johnson_Y_matrices_file = os.path.join(data_dir,'Y-Kinases.xlsx')\n",
    "densitometry_file = os.path.join(data_dir,'ST-Kinases_densitometry.xlsx')\n",
    "\n",
    "ST_matrices = PWM_Matrices(johnson_ST_matrices_file)\n",
    "ST_matrices.add_densitometry(densitometry_file)\n",
    "\n",
    "Y_matrices = PWM_Matrices(johnson_Y_matrices_file)\n",
    "\n",
    "st_scoring = Scoring(ST_matrices)\n",
    "y_scoring = Scoring(Y_matrices)\n",
    "\n",
    "ochoa_background_file = os.path.join(data_dir, 'johnson_ochoa_background_wfav.tsv')\n",
    "\n",
    "tyrosine_background_file = os.path.join(data_dir, 'johnson_tyrosine_background.tsv')\n",
    "\n",
    "ST_Background = PeptideBackground(ochoa_background_file)\n",
    "Y_Background = PeptideBackground(tyrosine_background_file)\n",
    "\n",
    "human_kinase_file = os.path.join(data_dir,'human_kinases_final.tsv')\n",
    "\n",
    "ortholog_manager = OrthologManager('orthologs', human_kinase_file=human_kinase_file)\n",
    "\n",
    "background = {'ST':ST_Background, 'Y':Y_Background}\n",
    "scoring = {'ST':st_scoring, 'Y':y_scoring}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Yeast Experiments\n",
    "\n",
    "### Leutert, M., Barente, A.S., Fukuda, N.K. et al. The regulatory landscape of the yeast phosphoproteome. Nat Struct Mol Biol 30, 1761–1773 (2023).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_phosphopeptides(phosphorylation_file, fasta_file, output_file) :\n",
    "    exp_df = pd.read_csv(phosphorylation_file)\n",
    "\n",
    "    #read in the fasta file\n",
    "    fasta_dict = seqio.to_dict(seqio.parse(fasta_file, 'fasta'))\n",
    "\n",
    "    exp_df = exp_df[exp_df['p_residue'].isin(['S', 'T', 'Y'])]\n",
    "\n",
    "    systematic_name_to_SGD_dict = {name:record.description.split(', ')[0].split(' ')[2] for name,record in fasta_dict.items()}\n",
    "    #systematic_name_to_SGD_dict = {name:SGD.split(':')[1] for name,SGD in systematic_name_to_SGD_dict.items()}\n",
    "\n",
    "    exp_df['SGDID'] = exp_df['systematic_name'].map(systematic_name_to_SGD_dict)\n",
    "\n",
    "    systematic_name_to_seq_dict = {name:str(record.seq) for name,record in fasta_dict.items()}\n",
    "    exp_df['sequence'] = (exp_df['systematic_name'].map(systematic_name_to_seq_dict))\n",
    "\n",
    "    #remove rows with missing sequences\n",
    "    exp_df = exp_df[~exp_df['sequence'].isna()]\n",
    "\n",
    "    #remove rows with stop codons\n",
    "    exp_df['sequence'] = exp_df['sequence'].str.rstrip('*')\n",
    "\n",
    "    #get sequence lengths\n",
    "    exp_df['seq_len'] = exp_df['sequence'].str.len()\n",
    "\n",
    "    #keep only rows where the sequence length is greater than or equal to the position of the phosphorylation site\n",
    "    exp_df = exp_df[exp_df['seq_len'] >= exp_df['p_position']]\n",
    "\n",
    "    exp_df['sanity_check'] = exp_df.apply(lambda row: row['sequence'][row['p_position']-1] == row['p_residue'], axis=1)\n",
    "    \n",
    "    exp_df = exp_df[exp_df['sanity_check']]\n",
    "\n",
    "    #if p_position is less than 6, add 5 - p_position number of _'s to the beginning of the sequence\n",
    "    exp_df['sequence'] = exp_df.apply(lambda row: '_'*(6 - row['p_position']) + row['sequence'], axis=1)\n",
    "\n",
    "    #if p_position is greater than seq_len - 6, add p_position - (seq_len - 5) number of _'s to the end of the sequence\n",
    "    exp_df['sequence'] = exp_df.apply(lambda row: row['sequence'] + '_'*(row['p_position'] - (row['seq_len'] - 5)), axis=1)\n",
    "\n",
    "    #make window around p_position if p_position is greater than 5 from p_position - 6 to p_position + 5, else from 0 to 10\n",
    "    exp_df['window'] = exp_df.apply(lambda row: row['sequence'][row['p_position']-6:row['p_position']+5] if row['p_position'] > 5 else row['sequence'][0:11], axis=1)\n",
    "\n",
    "    exp_df['adj_p_value'] = exp_df['adj_p_value'].apply(lambda x: -1*np.log10(x))\n",
    "\n",
    "    exp_df_filtered = exp_df[['SGDID', 'p_position', 'window', 'fc_log2', 'adj_p_value']].copy()\n",
    "\n",
    "    exp_df_filtered.rename(columns={'SGDID':'SGD', 'p_position':'site', 'window':'peptide', 'fc_log2':'log2fc', 'adj_p_value':'adjpvalue'}, inplace=True)\n",
    "\n",
    "    exp_df_filtered.to_csv(output_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5258\n",
      "Elapsed time for percentiles and matches: 6.01 seconds\n"
     ]
    }
   ],
   "source": [
    "published_yeast_file = 'yeast-perturbation.xlsx'\n",
    "published_yeast_path = os.path.join(test_dir, published_yeast_file)\n",
    "\n",
    "\n",
    "import gzip\n",
    "yeast_proteome_path = os.path.join(proteomes_dir, 'yeast_proteome.fasta')\n",
    "\n",
    "if not os.path.exists(proteomes_dir):\n",
    "    os.makedirs(proteomes_dir)\n",
    "    \n",
    "if not os.path.exists(yeast_proteome_path):\n",
    "    print('Downloading yeast proteome')\n",
    "    Utility.download_file('http://sgd-archive.yeastgenome.org/sequence/S288C_reference/orf_protein/orf_trans.fasta.gz', 'temp.gz')\n",
    "    with gzip.open('temp.gz', 'rb') as f_in:\n",
    "        with open(yeast_proteome_path, 'wb') as f_out:\n",
    "            f_out.write(f_in.read())\n",
    "    os.remove('temp.gz')\n",
    "    \n",
    "\n",
    "if not os.path.exists(published_yeast_path):\n",
    "    print('Downloading yeast perturbation data')\n",
    "    Utility.download_file('https://static-content.springer.com/esm/art%3A10.1038%2Fs41594-023-01115-3/MediaObjects/41594_2023_1115_MOESM9_ESM.xlsx',\n",
    "                  published_yeast_path)\n",
    "\n",
    "\n",
    "HOG1_experiment_file = os.path.join(test_dir,'HOG1_exp.csv')\n",
    "\n",
    "if not os.path.exists(HOG1_experiment_file):\n",
    "    df = pd.read_excel(published_yeast_path, sheet_name='p_site_diff_reg')\n",
    "\n",
    "HOG1_phosphopeptides_file = 'yeast_HOG1_KC_phosphopeptides.csv'\n",
    "HOG1_path = os.path.join(test_dir, HOG1_phosphopeptides_file)\n",
    "if not os.path.exists(HOG1_path):\n",
    "    df[df['treatment_id'] == 'KC'].to_csv(HOG1_path, index=False)\n",
    "    clean_phosphopeptides(HOG1_path, yeast_proteome_path, HOG1_experiment_file)\n",
    "    \n",
    "column_names_dict={'id':'SGD', 'site':'site', 'peptide':'peptide', 'log2fc':'log2fc', 'dependent':'adjpvalue'}\n",
    "\n",
    "HOG1_df = pd.read_csv(HOG1_experiment_file)\n",
    "print(len(HOG1_df))\n",
    "session = Session(0, 'yeast', HOG1_df, column_names_dict, scoring, background, ortholog_manager, id_type='SGD', ambiguous=True, debug=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNF1_experiment_file = os.path.join(test_dir,'SNF1_exp.csv')\n",
    "\n",
    "if not os.path.exists(SNF1_experiment_file):\n",
    "    df = pd.read_excel(published_yeast_path, sheet_name='p_site_diff_reg')\n",
    "\n",
    "SNF1_phosphopeptides_file = 'yeast_SNF1_GL_phosphopeptides.csv'\n",
    "SNF1_path = os.path.join(test_dir, SNF1_phosphopeptides_file)\n",
    "if not os.path.exists(SNF1_path):\n",
    "    df[df['treatment_id'] == 'GL'].to_csv(SNF1_path, index=False)\n",
    "    clean_phosphopeptides(SNF1_path, yeast_proteome_path, SNF1_experiment_file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mouse Experiment\n",
    "### Huttlin EL, Jedrychowski MP, Elias JE, Goswami T, Rad R, Beausoleil SA, Villén J, Haas W, Sowa ME, Gygi SP. A tissue-specific atlas of mouse protein phosphorylation and expression. Cell. 2010 Dec 23;143(7):1174-89. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def region_check(region) :\n",
    "    regions = [int(r) for r in region.split(';')]\n",
    "    if regions[0] != regions[1] :\n",
    "        return False\n",
    "    else :\n",
    "        return True\n",
    "\n",
    "def extract_MGI_id(gene_ids) :\n",
    "    if type(gene_ids) != str :\n",
    "        return None\n",
    "    if(len(gene_ids) == 0) :\n",
    "        return None\n",
    "    ids = gene_ids.split('; ')\n",
    "    \n",
    "    for id in ids :\n",
    "        if 'MGI' in id :\n",
    "            #expected format MG:MGI:1234567\n",
    "            mgi = id.split(':')[2]\n",
    "            return mgi\n",
    "    else :\n",
    "        return None\n",
    "    \n",
    "#function to process peptides\n",
    "def extract_peptide(opeptide) :\n",
    "    #split peptide by *\n",
    "    peptide = opeptide.split('*')\n",
    "\n",
    "    #if there are not 2 parts, return None\n",
    "    if len(peptide) != 2 :\n",
    "        print(opeptide)\n",
    "        return None\n",
    "    #get last 5 elements from first part\n",
    "    first_part = peptide[0][-6:]\n",
    "\n",
    "    #get first 5 elements from second part\n",
    "    second_part = peptide[1][:5]\n",
    "\n",
    "    #combine\n",
    "    return first_part + second_part"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>log2fc</th>\n",
       "      <th>peptide</th>\n",
       "      <th>site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EG:72925; EN:ENSMUSG00000036469; MG:MGI:192017...</td>\n",
       "      <td>1.611158</td>\n",
       "      <td>SNISKAS*SPTTGT</td>\n",
       "      <td>52;52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EG:72925; EN:ENSMUSG00000036469; MG:MGI:192017...</td>\n",
       "      <td>1.611158</td>\n",
       "      <td>NISKASS*PTTGTA</td>\n",
       "      <td>53;53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EG:57438; EN:ENSMUSG00000026977; MG:MGI:193105...</td>\n",
       "      <td>1.611158</td>\n",
       "      <td>MEDLETS*EDEF--</td>\n",
       "      <td>689;689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EG:71779; EN:ENSMUSG00000025702; MG:MGI:191902...</td>\n",
       "      <td>1.389903</td>\n",
       "      <td>SAFSRTS*VTPSNQ</td>\n",
       "      <td>66;67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EG:71779; EN:ENSMUSG00000025702; MG:MGI:191902...</td>\n",
       "      <td>1.389903</td>\n",
       "      <td>RTSVTPS*NQDICR</td>\n",
       "      <td>71;71</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  id    log2fc  \\\n",
       "0  EG:72925; EN:ENSMUSG00000036469; MG:MGI:192017...  1.611158   \n",
       "1  EG:72925; EN:ENSMUSG00000036469; MG:MGI:192017...  1.611158   \n",
       "2  EG:57438; EN:ENSMUSG00000026977; MG:MGI:193105...  1.611158   \n",
       "3  EG:71779; EN:ENSMUSG00000025702; MG:MGI:191902...  1.389903   \n",
       "4  EG:71779; EN:ENSMUSG00000025702; MG:MGI:191902...  1.389903   \n",
       "\n",
       "          peptide     site  \n",
       "0  SNISKAS*SPTTGT    52;52  \n",
       "1  NISKASS*PTTGTA    53;53  \n",
       "2  MEDLETS*EDEF--  689;689  \n",
       "3  SAFSRTS*VTPSNQ    66;67  \n",
       "4  RTSVTPS*NQDICR    71;71  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "published_mouse_data = 'mouse-phosphoproteome.zip'\n",
    "mouse_phosphopeptides_file = os.path.join(test_dir,'cell5459mmc2.xlsb')\n",
    "\n",
    "if not os.path.exists(mouse_phosphopeptides_file):\n",
    "    print('Downloading mouse phosphoproteome data')\n",
    "    download_file('https://ars.els-cdn.com/content/image/1-s2.0-S0092867410013796-mmc2.zip',\n",
    "                  published_mouse_data)\n",
    "    #extract the zip file into the current directory\n",
    "    with zipfile.ZipFile(published_mouse_data, 'r') as zip_ref:\n",
    "        zip_ref.extractall(test_dir)\n",
    "        #delete the zip file\n",
    "        os.remove(published_mouse_data)\n",
    "\n",
    "df = pd.read_excel(mouse_phosphopeptides_file, engine='pyxlsb', sheet_name='phosphoMouse_siteList_withAllPr')\n",
    "df = df[[\"Gene ID's\", 'entropy', 'motif peptide', 'Region']].copy()\n",
    "\n",
    "#rename columns to id, log2fc, peptide, site\n",
    "df.columns = ['id', 'log2fc', 'peptide', 'site']\n",
    "\n",
    "display(df.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QQICRDSEDEEAW\n",
      "29978\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>log2fc</th>\n",
       "      <th>peptide</th>\n",
       "      <th>site</th>\n",
       "      <th>region_check</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1920175</td>\n",
       "      <td>1.611158</td>\n",
       "      <td>NISKASSPTTG</td>\n",
       "      <td>52</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1920175</td>\n",
       "      <td>1.611158</td>\n",
       "      <td>ISKASSPTTGT</td>\n",
       "      <td>53</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1931053</td>\n",
       "      <td>1.611158</td>\n",
       "      <td>EDLETSEDEF-</td>\n",
       "      <td>689</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1919029</td>\n",
       "      <td>1.389903</td>\n",
       "      <td>TSVTPSNQDIC</td>\n",
       "      <td>71</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1858916</td>\n",
       "      <td>0.982129</td>\n",
       "      <td>FPECDSDEDEE</td>\n",
       "      <td>206</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35959</th>\n",
       "      <td>2444286</td>\n",
       "      <td>1.664698</td>\n",
       "      <td>RLRLLSFRSME</td>\n",
       "      <td>1535</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35960</th>\n",
       "      <td>2444286</td>\n",
       "      <td>1.757968</td>\n",
       "      <td>LLSFRSMEETR</td>\n",
       "      <td>1538</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35961</th>\n",
       "      <td>2444286</td>\n",
       "      <td>1.576964</td>\n",
       "      <td>DRPVSSPGEAE</td>\n",
       "      <td>2407</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35962</th>\n",
       "      <td>1920453</td>\n",
       "      <td>1.590218</td>\n",
       "      <td>RDSWVSPRKRR</td>\n",
       "      <td>89</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35963</th>\n",
       "      <td>1920453</td>\n",
       "      <td>1.611158</td>\n",
       "      <td>RKRRLSSSEKD</td>\n",
       "      <td>96</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29978 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            id    log2fc      peptide  site  region_check\n",
       "0      1920175  1.611158  NISKASSPTTG    52          True\n",
       "1      1920175  1.611158  ISKASSPTTGT    53          True\n",
       "2      1931053  1.611158  EDLETSEDEF-   689          True\n",
       "4      1919029  1.389903  TSVTPSNQDIC    71          True\n",
       "5      1858916  0.982129  FPECDSDEDEE   206          True\n",
       "...        ...       ...          ...   ...           ...\n",
       "35959  2444286  1.664698  RLRLLSFRSME  1535          True\n",
       "35960  2444286  1.757968  LLSFRSMEETR  1538          True\n",
       "35961  2444286  1.576964  DRPVSSPGEAE  2407          True\n",
       "35962  1920453  1.590218  RDSWVSPRKRR    89          True\n",
       "35963  1920453  1.611158  RKRRLSSSEKD    96          True\n",
       "\n",
       "[29978 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df['region_check'] = df['site'].apply(region_check)\n",
    "df['site'] = df['site'].apply(lambda x: x.split(';')[0])\n",
    "\n",
    "#remove rows where region is not the same\n",
    "df = df[df['region_check']].copy()\n",
    "\n",
    "df['id'] = df['id'].apply(extract_MGI_id)\n",
    "\n",
    "#remove rows where id is None\n",
    "df = df[df['id'].notna()].copy()\n",
    "\n",
    "df['peptide'] = df['peptide'].apply(extract_peptide)\n",
    "\n",
    "#remove rows where peptide is None\n",
    "df = df[df['peptide'].notna()].copy()\n",
    "\n",
    "#remove duplicate rows\n",
    "df = df.drop_duplicates().copy()\n",
    "\n",
    "#how many rows are left?\n",
    "print(len(df))\n",
    "\n",
    "#write to csv\n",
    "mouse_experiment_file = os.path.join(test_dir,'mouse_complete_phosphopeptides.csv')\n",
    "df.to_csv(mouse_experiment_file, index=False)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Duplicated IDs in ST data: 633      2442238_220\n",
      "634      2442238_223\n",
      "964      1351658_194\n",
      "1225     1890410_263\n",
      "1226     1890410_267\n",
      "            ...     \n",
      "26834      1914395_8\n",
      "26921     107749_136\n",
      "26938     1913564_96\n",
      "26956      98809_283\n",
      "27900     894652_359\n",
      "Name: __ENTRY_ID, Length: 102, dtype: object\n",
      "Elapsed time for percentiles and matches: 92.41 seconds\n"
     ]
    }
   ],
   "source": [
    "\n",
    "column_names_dict={'id':'id', 'site':'site', 'peptide':'peptide', 'log2fc':'log2fc', 'dependent':None}\n",
    "\n",
    "session = Session(0, 'mouse', df, column_names_dict, scoring, background, ortholog_manager, id_type='MGI', ambiguous=True, debug=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fly Experiment\n",
    "\n",
    "### Yang L & Zhu AW et al. ERK controls embryonic cleavage divisions in Drosophila. Developmental Cell (2024)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4139\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Motif</th>\n",
       "      <th>log2FC</th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>Site Position</th>\n",
       "      <th>neglogadjp</th>\n",
       "      <th>adjp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KWRDESDDEE</td>\n",
       "      <td>-0.000618</td>\n",
       "      <td>Q9VH95</td>\n",
       "      <td>116</td>\n",
       "      <td>0.003350</td>\n",
       "      <td>0.992315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EGGKESPPPP</td>\n",
       "      <td>0.001109</td>\n",
       "      <td>Q9VGU5</td>\n",
       "      <td>589</td>\n",
       "      <td>0.003619</td>\n",
       "      <td>0.991701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GKNDDSDEEE</td>\n",
       "      <td>-0.000843</td>\n",
       "      <td>Q9VZP5</td>\n",
       "      <td>76</td>\n",
       "      <td>0.003619</td>\n",
       "      <td>0.991701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KLRKVSLDEP</td>\n",
       "      <td>-0.001183</td>\n",
       "      <td>A1Z898</td>\n",
       "      <td>424</td>\n",
       "      <td>0.003804</td>\n",
       "      <td>0.991280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LHDEASDDKS</td>\n",
       "      <td>0.003086</td>\n",
       "      <td>Q8IMV6</td>\n",
       "      <td>205</td>\n",
       "      <td>0.006157</td>\n",
       "      <td>0.985923</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Motif    log2FC Uniprot  Site Position  neglogadjp      adjp\n",
       "0  KWRDESDDEE -0.000618  Q9VH95            116    0.003350  0.992315\n",
       "1  EGGKESPPPP  0.001109  Q9VGU5            589    0.003619  0.991701\n",
       "2  GKNDDSDEEE -0.000843  Q9VZP5             76    0.003619  0.991701\n",
       "3  KLRKVSLDEP -0.001183  A1Z898            424    0.003804  0.991280\n",
       "4  LHDEASDDKS  0.003086  Q8IMV6            205    0.006157  0.985923"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "published_fly_data = os.path.join(test_dir,'optosos_fly_clean.csv')\n",
    "df = pd.read_csv(published_fly_data)\n",
    "print(len(df))\n",
    "display(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Elapsed time for percentiles and matches: 7.98 seconds\n"
     ]
    }
   ],
   "source": [
    "column_names_dict={'id':'Uniprot', 'site':'Site Position', 'peptide':'Motif', 'log2fc':'log2FC', 'dependent': 'neglogadjp'}\n",
    "session = Session(0, 'fly', df, column_names_dict, scoring, background, ortholog_manager, id_type='UniProtKB', ambiguous=True, debug=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Zebrafish experiment\n",
    "\n",
    "### Kwon, O.K., Kim, S.J., Lee, Y.-M., Lee, Y.-H., Bae, Y.-S., Kim, J.Y., Peng, X., Cheng, Z., Zhao, Y. and Lee, S. (2016), Global analysis of phosphoproteome dynamics in embryonic development of zebrafish (Danio rerio). Proteomics, 16: 136-149"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "published_zebrafish_data = os.path.join(test_dir,'Kwon_zebrafish.xls')\n",
    "final_zebrafish_data = os.path.join(test_dir,'zebrafish_phosphopeptides.csv')\n",
    "\n",
    "proteome_url = 'https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000000437/UP000000437_7955.fasta.gz'\n",
    "\n",
    "if not os.path.exists(final_zebrafish_data):\n",
    "    print('Downloading zebrafish phosphoproteome data')\n",
    "    #download_file('https://analyticalsciencejournals.onlinelibrary.wiley.com/action/downloadSupplement?doi=10.1002%2Fpmic.201500017&file=pmic12168-sup-0003-tableS3.xls',\n",
    "    #              'Kwon_zebrafish.xls')\n",
    "    df = pd.read_excel(published_zebrafish_data, sheet_name='phosphorylation site', skiprows=2)\n",
    "    df.to_csv(final_zebrafish_data, index=False)\n",
    "\n",
    "proteome_fasta = os.path.join(proteomes_dir,'zebrafish_proteome.fasta')\n",
    "\n",
    "if not os.path.exists(proteome_fasta):\n",
    "    print('Downloading zebrafish proteome')\n",
    "    download_file(proteome_url, 'zebrafish_proteome.fasta.gz')\n",
    "    os.system('gunzip zebrafish_proteome.fasta.gz')\n",
    "    os.rename('zebrafish_proteome.fasta', proteome_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tage</th>\n",
       "      <th>UniProtKB ID</th>\n",
       "      <th>Phospho Position</th>\n",
       "      <th>Phospo Amino</th>\n",
       "      <th>Motif Logo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0JMB8-889</td>\n",
       "      <td>A0JMB8</td>\n",
       "      <td>889</td>\n",
       "      <td>S</td>\n",
       "      <td>******S**D***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0MQ61-1201</td>\n",
       "      <td>A0MQ61</td>\n",
       "      <td>1201</td>\n",
       "      <td>S</td>\n",
       "      <td>******S**D***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1L1W1-687</td>\n",
       "      <td>A1L1W1</td>\n",
       "      <td>687</td>\n",
       "      <td>S</td>\n",
       "      <td>******S**D***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1L260-299</td>\n",
       "      <td>A1L260</td>\n",
       "      <td>299</td>\n",
       "      <td>S</td>\n",
       "      <td>******S**D***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2RRV2-60</td>\n",
       "      <td>A2RRV2</td>\n",
       "      <td>60</td>\n",
       "      <td>S</td>\n",
       "      <td>******S**D***</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Tage UniProtKB ID  Phospho Position Phospo Amino     Motif Logo\n",
       "0   A0JMB8-889       A0JMB8               889            S  ******S**D***\n",
       "1  A0MQ61-1201       A0MQ61              1201            S  ******S**D***\n",
       "2   A1L1W1-687       A1L1W1               687            S  ******S**D***\n",
       "3   A1L260-299       A1L260               299            S  ******S**D***\n",
       "4    A2RRV2-60       A2RRV2                60            S  ******S**D***"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of proteins in zebrafish proteome: 25999\n",
      "Number of proteins in zebrafish proteome that have phosphosites: 1078\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(final_zebrafish_data)\n",
    "display(df.head())\n",
    "\n",
    "zebrafish_fasta_dict = seqio.to_dict(seqio.parse(proteome_fasta, 'fasta'), key_function=lambda x: x.id.split('|')[1])\n",
    "print(f'Number of proteins in zebrafish proteome: {len(zebrafish_fasta_dict)}')\n",
    "\n",
    "df = df[df['Phospo Amino'].isin(['S', 'T', 'Y'])]\n",
    "df['exists_in_proteome'] = df['UniProtKB ID'].apply(lambda x: x in zebrafish_fasta_dict)\n",
    "df = df[df['exists_in_proteome']]\n",
    "\n",
    "print(f'Number of proteins in zebrafish proteome that have phosphosites: {len(df)}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GQPGFKFTVAE\n",
      "RSRGEAEDSSS\n",
      "HKSKSEEVGTV\n",
      "TSCETKKESLR\n",
      "EAPATRRSSRR\n",
      "PAITSDPEQDE\n",
      "ARRLLGADNAA\n",
      "KSRGKLRRKVH\n",
      "FKNEQEVVSDD\n",
      "HQRRALTPSAS\n",
      "VLPLQPLIEKP\n",
      "PVTRPHTTKRG\n",
      "LESGSKVTEAC\n",
      "NADIPNEPESP\n",
      "GGEYRPKVYDQ\n",
      "NSVSSDLKGSP\n",
      "QLSPSRENETD\n",
      "TPAVKNLDMSS\n",
      "KKNQEKSLSDR\n",
      "PGPDGPMGGMG\n",
      "PNNMAGMNNPP\n",
      "VQQQLQQKQQQ\n",
      "VDLIPEPDTTE\n",
      "AEEEEVVTKNV\n",
      "ISKMAAISKAG\n",
      "AAFAEKNRGRV\n",
      "RGRVMGGLPDV\n",
      "WSPSHQRRALT\n",
      "KSLSDRSKANK\n",
      "TSRKPEHRDDT\n",
      "PQSPSKSTSSN\n",
      "AKASKAPVSPH\n",
      "TEMQRHYVMYY\n",
      "KAAAFAEKNRG\n",
      "RIKDEFQFLQA\n",
      "PKIRQRIPFDL\n",
      "WSWAVQWLQKK\n",
      "LRPVGDSSVGR\n",
      "APHQPGQPGFK\n",
      "VPVDAPNPLTS\n",
      "MRVLLHHQQTD\n",
      "KATAEQDNIQV\n",
      "IRSLFVCRHIK\n",
      "ALQEALKEKQQ\n",
      "PEAVKRKSSTD\n",
      "KDAAKEASASF\n",
      "SVKKGERELKI\n",
      "GERELKINDRV\n",
      "EGLPNHVHQGQ\n",
      "RLRQQLKVPDK\n",
      "VIRSLFVCRHI\n",
      "RPPTPPPVEEV\n",
      "SRRSRGEAEDS\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peptide</th>\n",
       "      <th>UniProtKB ID</th>\n",
       "      <th>site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRKADSGKDLL</td>\n",
       "      <td>A0JMB8</td>\n",
       "      <td>889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PKGRKSSPDVT</td>\n",
       "      <td>A1L260</td>\n",
       "      <td>299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>RGLSASNVDLT</td>\n",
       "      <td>A4IG65</td>\n",
       "      <td>606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TEKEDSDSDDV</td>\n",
       "      <td>A5PLK1</td>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>IPLNKSHNDFV</td>\n",
       "      <td>B3DLH3</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3488</th>\n",
       "      <td>ETKRISVTQIP</td>\n",
       "      <td>Q8JG61</td>\n",
       "      <td>830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3489</th>\n",
       "      <td>LLDAESTRGDD</td>\n",
       "      <td>Q8UVJ6</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3490</th>\n",
       "      <td>_MKGDTPVNST</td>\n",
       "      <td>Q90Z40</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3496</th>\n",
       "      <td>LGDDESSSVST</td>\n",
       "      <td>Q9W6A5</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3498</th>\n",
       "      <td>EPNVSYICSRY</td>\n",
       "      <td>Q9YH61</td>\n",
       "      <td>243</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1025 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          peptide UniProtKB ID  site\n",
       "0     PRKADSGKDLL       A0JMB8   889\n",
       "3     PKGRKSSPDVT       A1L260   299\n",
       "6     RGLSASNVDLT       A4IG65   606\n",
       "7     TEKEDSDSDDV       A5PLK1   177\n",
       "10    IPLNKSHNDFV       B3DLH3   106\n",
       "...           ...          ...   ...\n",
       "3488  ETKRISVTQIP       Q8JG61   830\n",
       "3489  LLDAESTRGDD       Q8UVJ6    59\n",
       "3490  _MKGDTPVNST       Q90Z40     5\n",
       "3496  LGDDESSSVST       Q9W6A5   333\n",
       "3498  EPNVSYICSRY       Q9YH61   243\n",
       "\n",
       "[1025 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df['peptide'] = df.apply(lambda x: get_protein_sequence(str(zebrafish_fasta_dict[x['UniProtKB ID']].seq), x['Phospho Position'] - 1 ), axis=1)\n",
    "df = df.rename(columns={'Phospho Position': 'site'})\n",
    "\n",
    "df = df[['peptide', 'UniProtKB ID', 'site',]]\n",
    "#drop rows with missing peptides\n",
    "df = df.dropna(subset=['peptide'])\n",
    "\n",
    "zebrafish_experiment_file = os.path.join(test_dir,'zebrafish_experiment.csv')\n",
    "df.to_csv(zebrafish_experiment_file, index=False)\n",
    "\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Elapsed time for percentiles and matches: 3.32 seconds\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(zebrafish_experiment_file)\n",
    "column_names_dict={'id':'UniProtKB ID', 'site':'site', 'peptide':'peptide', 'log2fc':None, 'dependent': None}\n",
    "\n",
    "session = Session(0, 'zebrafish', df, column_names_dict, scoring, background, ortholog_manager, id_type='UniProtKB', ambiguous=True, debug=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Worm experiment\n",
    "\n",
    "### Li WJ, Wang CW, Tao L, Yan YH, Zhang MJ, Liu ZX, Li YX, Zhao HQ, Li XM, He XD, Xue Y, Dong MQ. Insulin signaling regulates longevity through protein phosphorylation in Caenorhabditis elegans. Nat Commun. 2021 Jul 27;12(1):4568."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phosphopeptide (phosphorylation marked as * at the right side of S/T/Y residues)</th>\n",
       "      <th>Charge state</th>\n",
       "      <th>UniProt ID</th>\n",
       "      <th>Position</th>\n",
       "      <th>Residue</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Gene</th>\n",
       "      <th>WT_Bio1</th>\n",
       "      <th>WT_Bio2</th>\n",
       "      <th>WT_Bio3</th>\n",
       "      <th>...</th>\n",
       "      <th>daf-2_Bio3</th>\n",
       "      <th>daf-2_Bio4</th>\n",
       "      <th>daf-16; daf-2_Bio1</th>\n",
       "      <th>daf-16; daf-2_Bio2</th>\n",
       "      <th>daf-16; daf-2_Bio3</th>\n",
       "      <th>daf-16; daf-2_Bio4</th>\n",
       "      <th>daf-16_Bio1</th>\n",
       "      <th>daf-16_Bio2</th>\n",
       "      <th>daf-16_Bio3</th>\n",
       "      <th>daf-16_Bio4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>APGNQAQVRKLS*QVMQHGR</td>\n",
       "      <td>2</td>\n",
       "      <td>A4F337</td>\n",
       "      <td>75</td>\n",
       "      <td>S</td>\n",
       "      <td>Protein 2RSSE.1, isoform a</td>\n",
       "      <td>2RSSE.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAFASNPS*HSLDYQEVGASNPR</td>\n",
       "      <td>3</td>\n",
       "      <td>U4PAK0</td>\n",
       "      <td>427</td>\n",
       "      <td>S</td>\n",
       "      <td>Protein 2RSSE.1, isoform b</td>\n",
       "      <td>2RSSE.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.689184</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAFASNPSHS*LDYQEVGASNPR</td>\n",
       "      <td>3</td>\n",
       "      <td>U4PAK0</td>\n",
       "      <td>429</td>\n",
       "      <td>S</td>\n",
       "      <td>Protein 2RSSE.1, isoform b</td>\n",
       "      <td>2RSSE.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.320352</td>\n",
       "      <td>8.408425</td>\n",
       "      <td>...</td>\n",
       "      <td>1.946737</td>\n",
       "      <td>6.297725</td>\n",
       "      <td>10.161053</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.672665</td>\n",
       "      <td>18.279308</td>\n",
       "      <td>10.756736</td>\n",
       "      <td>11.226915</td>\n",
       "      <td>11.707134</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAFASNPSHS*LDYQEVGASNPR</td>\n",
       "      <td>2</td>\n",
       "      <td>U4PAK0</td>\n",
       "      <td>429</td>\n",
       "      <td>S</td>\n",
       "      <td>Protein 2RSSE.1, isoform b</td>\n",
       "      <td>2RSSE.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.387018</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>14.048561</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GHTS*VEDDTWLAEVVPHDEIPK</td>\n",
       "      <td>3</td>\n",
       "      <td>U4PAK0</td>\n",
       "      <td>447</td>\n",
       "      <td>S</td>\n",
       "      <td>Protein 2RSSE.1, isoform b</td>\n",
       "      <td>2RSSE.1</td>\n",
       "      <td>1.455456</td>\n",
       "      <td>7.980528</td>\n",
       "      <td>5.605617</td>\n",
       "      <td>...</td>\n",
       "      <td>17.520635</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Phosphopeptide (phosphorylation marked as * at the right side of S/T/Y residues)  \\\n",
       "0                               APGNQAQVRKLS*QVMQHGR                                 \n",
       "1                           AAAFASNPS*HSLDYQEVGASNPR                                 \n",
       "2                           AAAFASNPSHS*LDYQEVGASNPR                                 \n",
       "3                           AAAFASNPSHS*LDYQEVGASNPR                                 \n",
       "4                            GHTS*VEDDTWLAEVVPHDEIPK                                 \n",
       "\n",
       "   Charge state UniProt ID  Position Residue                     Protein  \\\n",
       "0             2     A4F337        75       S  Protein 2RSSE.1, isoform a   \n",
       "1             3     U4PAK0       427       S  Protein 2RSSE.1, isoform b   \n",
       "2             3     U4PAK0       429       S  Protein 2RSSE.1, isoform b   \n",
       "3             2     U4PAK0       429       S  Protein 2RSSE.1, isoform b   \n",
       "4             3     U4PAK0       447       S  Protein 2RSSE.1, isoform b   \n",
       "\n",
       "      Gene   WT_Bio1   WT_Bio2   WT_Bio3  ...  daf-2_Bio3  daf-2_Bio4  \\\n",
       "0  2RSSE.1  0.000000  0.000000  0.000000  ...    0.000000    0.000000   \n",
       "1  2RSSE.1  0.000000  0.000000  0.000000  ...    0.000000    0.000000   \n",
       "2  2RSSE.1  0.000000  5.320352  8.408425  ...    1.946737    6.297725   \n",
       "3  2RSSE.1  0.000000  0.000000  0.000000  ...    0.000000    0.000000   \n",
       "4  2RSSE.1  1.455456  7.980528  5.605617  ...   17.520635    0.000000   \n",
       "\n",
       "   daf-16; daf-2_Bio1  daf-16; daf-2_Bio2  daf-16; daf-2_Bio3  \\\n",
       "0            0.000000                 0.0            0.000000   \n",
       "1            0.000000                 0.0            0.000000   \n",
       "2           10.161053                 0.0           12.672665   \n",
       "3            3.387018                 0.0            0.000000   \n",
       "4            0.000000                 0.0            0.000000   \n",
       "\n",
       "   daf-16; daf-2_Bio4  daf-16_Bio1  daf-16_Bio2  daf-16_Bio3  daf-16_Bio4  \n",
       "0            0.000000     0.000000     0.000000     0.000000          0.0  \n",
       "1            0.000000     2.689184     0.000000     0.000000          0.0  \n",
       "2           18.279308    10.756736    11.226915    11.707134          0.0  \n",
       "3            0.000000     0.000000     0.000000    14.048561          0.0  \n",
       "4            0.000000     0.000000     0.000000     0.000000          0.0  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniProt ID</th>\n",
       "      <th>Position</th>\n",
       "      <th>log2fc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A061ACM2</td>\n",
       "      <td>442</td>\n",
       "      <td>-0.917498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A0A061ACS9</td>\n",
       "      <td>987</td>\n",
       "      <td>-0.689919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A0A061ACS9</td>\n",
       "      <td>995</td>\n",
       "      <td>-0.087314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A0A061ACS9</td>\n",
       "      <td>1024</td>\n",
       "      <td>0.232822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A0A061ACS9</td>\n",
       "      <td>1028</td>\n",
       "      <td>-0.111178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>A0PD34</td>\n",
       "      <td>134</td>\n",
       "      <td>-1.549724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>A3FPJ7</td>\n",
       "      <td>34</td>\n",
       "      <td>0.395414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>A3KFD2</td>\n",
       "      <td>345</td>\n",
       "      <td>-0.330772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>A3KFD2</td>\n",
       "      <td>354</td>\n",
       "      <td>-0.681597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>A3QM97</td>\n",
       "      <td>683</td>\n",
       "      <td>0.032887</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    UniProt ID  Position    log2fc\n",
       "3   A0A061ACM2       442 -0.917498\n",
       "7   A0A061ACS9       987 -0.689919\n",
       "8   A0A061ACS9       995 -0.087314\n",
       "10  A0A061ACS9      1024  0.232822\n",
       "11  A0A061ACS9      1028 -0.111178\n",
       "50      A0PD34       134 -1.549724\n",
       "53      A3FPJ7        34  0.395414\n",
       "55      A3KFD2       345 -0.330772\n",
       "57      A3KFD2       354 -0.681597\n",
       "64      A3QM97       683  0.032887"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of proteins in worm proteome: 26703\n",
      "      UniProt ID  Position    log2fc  exists_in_proteome\n",
      "122       A7DT28       556 -0.775533               False\n",
      "123       A7DT28       686 -0.366023               False\n",
      "125       A7DT28       807 -1.661523               False\n",
      "129       A7DT28      1355 -0.292911               False\n",
      "130       A7DT28      1357 -0.292911               False\n",
      "...          ...       ...       ...                 ...\n",
      "15070     Q9XVY4       749 -0.131792               False\n",
      "15174     Q9XWG7       377  0.692752               False\n",
      "15374   Q9Y1J3-2       255 -0.039455               False\n",
      "15377     S6EZL6        11  0.143245               False\n",
      "15428     V6CJB8       573 -0.282021               False\n",
      "\n",
      "[261 rows x 4 columns]\n",
      "Number of valid phosphosites: 4184\n",
      "ETGGYITLGYR\n",
      "SRSGIDVRRND\n",
      "SSRVTRDEQTV\n",
      "EEEVEADYAPS\n",
      "AFSQSPHRDWS\n",
      "QSPHRDWSAST\n",
      "WSASTVYDRVY\n",
      "ASTVYDRVYNS\n",
      "STVYDRVYNSN\n",
      "STQITETITRT\n",
      "QITETITRTTH\n",
      "SVRTQEFAERI\n",
      "PIQGMDTPFDH\n",
      "EKRYFGDTLEK\n",
      "DKKPNVPKKRA\n",
      "AGPPPPVIEIS\n",
      "KDGKEGKPFDF\n",
      "IKLGKEDQLRV\n",
      "LSRVGQPSSLV\n",
      "GQPSSLVVEVS\n",
      "LADLIDVDAST\n",
      "LIDVDASTLKI\n",
      "IVDPIGDRAKI\n",
      "QHLMYLTGGVP\n",
      "AVQDGRIPMEI\n",
      "ENAGRVEEAAK\n",
      "KFKRLQKCLKK\n",
      "RHTQGMVEAQF\n",
      "PASLREKINQK\n",
      "SLREKINQKKE\n",
      "VDQAQRAVITP\n",
      "QQLASRQKDEE\n",
      "LQEELHQKNIG\n",
      "VFPMHGEDEED\n",
      "RIVDDVTMDQF\n",
      "KRQSKPEPIFE\n",
      "HLSAQPSKPVL\n",
      "KSNDNFQKTKP\n",
      "HSRKLRKQMPA\n",
      "EASKILEHYPV\n",
      "YPVTEQTSPST\n",
      "FKSKRKNKKKR\n",
      "SRPTERTLRGL\n",
      "LQRVYPVEHTL\n",
      "GSMSVADLGSL\n",
      "MSVADLGSLTR\n",
      "RLEDKIRLLQE\n",
      "EDKIRLLQEDL\n",
      "LQEDLESEREL\n",
      "ELSKLRKLLEE\n",
      "KAQNTIALLER\n",
      "QNFVDKSQRLT\n",
      "RVLMTAQNEAD\n",
      "ADNSEEPSFVS\n",
      "QLHNAIPPMVN\n",
      "AVNKDLESGGR\n",
      "KKSGGIWEFFS\n",
      "LPEGLDGTDLG\n",
      "DLGACEWVELR\n",
      "YLRKRHRLCMM\n",
      "NGFGGATNGGS\n",
      "FGGATNGGSPA\n",
      "AESRVAEEKIR\n",
      "VAEEKIRRDRV\n",
      "EKIRRDRVRAE\n",
      "RAAGAVTPSDP\n",
      "VERGSLYSLDY\n",
      "FLILRENLITA\n",
      "VYPTIRDDSMS\n",
      "YANPFPKKAYE\n",
      "TSDSDATETLT\n",
      "PSEPVVVRTPS\n",
      "RTPSPVPQPTP\n",
      "PSPVPQPTPPP\n",
      "PTPPPPPPPVR\n",
      "APSVFGGDDDL\n",
      "SIPLFIPDQDG\n",
      "GGALSVITEAS\n",
      "LSMPAGMLNKS\n",
      "NKSTQPDRPTM\n",
      "PNVDPNANWFK\n",
      "SNNKFNQMLSK\n",
      "EVSNSNNTMTL\n",
      "ARPMTPKAAST\n",
      "MTPKAASTCKT\n",
      "RGESFEEEEMT\n",
      "EYVDYQDSLPR\n",
      "EPTVVEESTAL\n",
      "DALNKLMAVRS\n",
      "NVVGRNQTSEE\n",
      "IPNEPLDIIPP\n",
      "SSSDVRRKRTA\n",
      "TRQQSNVADIP\n",
      "EAASAVEEAED\n",
      "SAVEEAEDEGR\n",
      "TQSLVRVEEET\n",
      "SLVRVEEETTT\n",
      "LEGLQDLIQQQ\n",
      "VSSSSMDDYSM\n",
      "DEDAKKKKSLR\n",
      "REKSRKADEES\n",
      "NNPQPASMYGH\n",
      "SMYGHRSVQSP\n",
      "VNGKANSITLD\n",
      "SLIQSNMSVLP\n",
      "SNMSVLPNFLH\n",
      "IDAERRERERK\n",
      "EEERREFELIE\n",
      "YGRGNNFDRNK\n",
      "EGWVNWDE___\n",
      "VSLPTRFAADF\n",
      "PIAAVAFSEDG\n",
      "PRDLKHLDQTS\n",
      "AVNGGGSTASS\n",
      "TPTSKKSVTST\n",
      "PGSISARSRTE\n",
      "GSISARSRTES\n",
      "PPPPPPPPPPP\n",
      "ISTGTKPPTES\n",
      "ESLYKREVAQM\n",
      "QRINNNKGYRP\n",
      "KYSNSGTRCKL\n",
      "TRCKLPIPDMF\n"
     ]
    }
   ],
   "source": [
    "published_worm_data_url = 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8316574/bin/41467_2021_24816_MOESM3_ESM.xlsx'\n",
    "published_worm_data = os.path.join(test_dir,'worm_phosphosites.xlsx')\n",
    "\n",
    "if not os.path.exists(published_worm_data):\n",
    "    print('Downloading worm phosphoproteome data')\n",
    "    download_file(published_worm_data_url, published_worm_data)\n",
    "\n",
    "df = pd.read_excel(published_worm_data, sheet_name='Identified phosphosites', skiprows=3)\n",
    "display(df.head())\n",
    "\n",
    "#proteome_url = 'https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000001940/UP000001940_6239.fasta.gz'\n",
    "proteome_url = 'https://rest.uniprot.org/uniprotkb/stream?format=fasta&query=%28proteome%3AUP000001940%29'\n",
    "proteome_fasta = os.path.join(test_dir,'worm_proteome.fasta')\n",
    "\n",
    "if not os.path.exists(proteome_fasta):\n",
    "    print('Downloading worm proteome')\n",
    "    download_file(proteome_url, proteome_fasta)\n",
    "    #os.system('gunzip worm_proteome.fasta.gz')\n",
    "\n",
    "df = df[df['Residue'].isin(['S', 'T', 'Y'])]\n",
    "\n",
    "df = df[['UniProt ID', 'Position', 'WT_Bio1', 'WT_Bio2', 'WT_Bio3', 'daf-2_Bio1', 'daf-2_Bio2', 'daf-2_Bio3', 'daf-2_Bio4']]\n",
    "df = df.groupby(['UniProt ID', 'Position']).max().reset_index()\n",
    "\n",
    "#create a column WT_Bio that is True if at least two of the WT_Bio1, WT_Bio2, WT_Bio3 columns have a value greater than 0\n",
    "df['WT_Bio_exist'] = ((df['WT_Bio1'] > 0).astype(int) + (df['WT_Bio2'] > 0).astype(int) + (df['WT_Bio3'] > 0).astype(int)) > 1\n",
    "df['daf-2_Bio_exist'] = ((df['daf-2_Bio1'] > 0).astype(int) + (df['daf-2_Bio2'] > 0).astype(int) + (df['daf-2_Bio3'] > 0).astype(int) + (df['daf-2_Bio4'] > 0).astype(int)) > 1\n",
    "\n",
    "#average the non-zero values of WT_Bio1, WT_Bio2, and WT_Bio3 and save result in a new column called WT_Bio\n",
    "df = df[(df[\"WT_Bio_exist\"]) & (df[\"daf-2_Bio_exist\"])]\n",
    "df['WT_Bio'] = df.apply(lambda x: [x['WT_Bio1'], x['WT_Bio2'], x['WT_Bio3']], axis=1)\n",
    "df['WT_Bio'] = df['WT_Bio'].apply(lambda x: np.mean([i for i in x if i > 0]))\n",
    "\n",
    "#average the non-zero values of daf-2_Bio1, daf-2_Bio2, daf-2_Bio3, and daf-2_Bio4 and save result in a new column called daf-2_Bio\n",
    "df['daf-2_Bio'] = df.apply(lambda x: [x['daf-2_Bio1'], x['daf-2_Bio2'], x['daf-2_Bio3'], x['daf-2_Bio4']], axis=1)\n",
    "df['daf-2_Bio'] = df['daf-2_Bio'].apply(lambda x: np.mean([i for i in x if i > 0]))\n",
    "\n",
    "#calculate the log2 fold change of daf-2_Bio over WT_Bio and save result in a new column called log2_fold_change\n",
    "df['log2fc'] = np.log2(df['daf-2_Bio'] / df['WT_Bio'])\n",
    "\n",
    "df = df[['UniProt ID','Position','log2fc']]\n",
    "\n",
    "display(df.head(10))\n",
    "\n",
    "worm_fasta_dict = seqio.to_dict(seqio.parse(proteome_fasta, 'fasta'), key_function=lambda x: x.id.split('|')[1])\n",
    "print(f'Number of proteins in worm proteome: {len(worm_fasta_dict)}')\n",
    "\n",
    "df['exists_in_proteome'] = df['UniProt ID'].apply(lambda x: x in worm_fasta_dict)\n",
    "#print the phosphosites that are not in the proteome\n",
    "print(df[~df['exists_in_proteome']])\n",
    "\n",
    "#keep only the phosphosites that are in the proteome\n",
    "df = df[df['exists_in_proteome']]\n",
    "\n",
    "df.drop(columns=['exists_in_proteome'], inplace=True)\n",
    "\n",
    "df['is_valid'] = df.apply(lambda x: x['Position'] < len(worm_fasta_dict[x['UniProt ID']].seq), axis=1)\n",
    "\n",
    "#keep only the phosphosites that are valid\n",
    "df = df[df['is_valid']]\n",
    "print(f'Number of valid phosphosites: {len(df)}')\n",
    "\n",
    "df.drop(columns=['is_valid'], inplace=True)\n",
    "\n",
    "df['peptide'] = df.apply(lambda x: get_protein_sequence(worm_fasta_dict[x['UniProt ID']].seq, x['Position'] - 1), axis=1)\n",
    "\n",
    "df = df.dropna(subset=['peptide'])\n",
    "\n",
    "\n",
    "worm_experiment_file = os.path.join(test_dir,'worm_experiment.csv')\n",
    "df.to_csv(worm_experiment_file, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4061\n",
      "Elapsed time for percentiles and matches: 6.28 seconds\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(worm_experiment_file)\n",
    "print(len(df))\n",
    "column_names_dict={'id':'UniProt ID', 'site':'Position', 'peptide':'peptide', 'log2fc':'log2fc', 'dependent': None}\n",
    "\n",
    "session = Session(0, 'worm', df, column_names_dict, scoring, background, ortholog_manager, id_type='UniProtKB', ambiguous=True, debug=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Human experiment\n",
    "### Humphrey SJ, Yang G, Yang P, Fazakerley DJ, Stöckli J, Yang JY, James DE. Dynamic adipocyte phosphoproteome reveals that Akt directly regulates mTORC2. Cell Metab. 2013 Jun 4;17(6):1009-1020."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Phosphorylated Amino Acid</th>\n",
       "      <th>IPI 3.68 Position in Protein</th>\n",
       "      <th>IPI</th>\n",
       "      <th>Site Localization Class</th>\n",
       "      <th>Localization Probability</th>\n",
       "      <th>Sequence Window</th>\n",
       "      <th>Inhibitor Exps Insulin/Starved Median (log2)</th>\n",
       "      <th>LY Screen Insulin Regulated Direction</th>\n",
       "      <th>LY Exps Insulin/Starved Median (log2)</th>\n",
       "      <th>...</th>\n",
       "      <th>Andromeda Search Score</th>\n",
       "      <th>Uniprot ID</th>\n",
       "      <th>Proteins IPI IDs</th>\n",
       "      <th>Positions</th>\n",
       "      <th>Leading Proteins</th>\n",
       "      <th>Protein Descriptions</th>\n",
       "      <th>Modified Sequence</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Protein Names</th>\n",
       "      <th>PEP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>39326</td>\n",
       "      <td>T</td>\n",
       "      <td>426</td>\n",
       "      <td>IPI00224626</td>\n",
       "      <td>Class I</td>\n",
       "      <td>0.999318</td>\n",
       "      <td>QQNSSRTLEKNKK</td>\n",
       "      <td>-1.41</td>\n",
       "      <td>Negative</td>\n",
       "      <td>-1.46</td>\n",
       "      <td>...</td>\n",
       "      <td>261.67</td>\n",
       "      <td>Q3UGE0;Q3UNN1;Q5DTS3;Q8C2A3;O55131</td>\n",
       "      <td>IPI00224626;IPI00874440</td>\n",
       "      <td>426</td>\n",
       "      <td>IPI00224626</td>\n",
       "      <td>cell division cycle 10 homolog;Septin-7</td>\n",
       "      <td>_IIEQQNSSRT(ph)IEK_</td>\n",
       "      <td>Sept7;mKIAA4020;Cdc10;Sept7</td>\n",
       "      <td>Putative uncharacterized protein;MKIAA4020 pro...</td>\n",
       "      <td>2.730000e-55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1600027N09Rik</td>\n",
       "      <td>S</td>\n",
       "      <td>191</td>\n",
       "      <td>IPI00720110</td>\n",
       "      <td>Class I</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>KVLTANSNPSSPS</td>\n",
       "      <td>-2.41</td>\n",
       "      <td>Negative</td>\n",
       "      <td>-2.74</td>\n",
       "      <td>...</td>\n",
       "      <td>195.38</td>\n",
       "      <td>Q14AR7;Q2NL54</td>\n",
       "      <td>IPI00720110</td>\n",
       "      <td>191</td>\n",
       "      <td>IPI00720110</td>\n",
       "      <td>RIKEN cDNA 1600027N09 gene</td>\n",
       "      <td>_VITANS(ph)NPSS(ph)PSAAK_</td>\n",
       "      <td>1600027N09Rik;mCG_6726;RP23-74K21.4;RP23-74K21...</td>\n",
       "      <td>MCG6726, isoform CRA_f;Novel protein (Possible...</td>\n",
       "      <td>1.960000e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1700080G18Rik</td>\n",
       "      <td>S</td>\n",
       "      <td>308</td>\n",
       "      <td>IPI00108279</td>\n",
       "      <td>Class I</td>\n",
       "      <td>0.998163</td>\n",
       "      <td>FKAPTRSVNGELW</td>\n",
       "      <td>-2.64</td>\n",
       "      <td>Negative</td>\n",
       "      <td>-2.27</td>\n",
       "      <td>...</td>\n",
       "      <td>70.98</td>\n",
       "      <td>Q9CVQ8</td>\n",
       "      <td>IPI00108279;IPI00808211</td>\n",
       "      <td>308;382</td>\n",
       "      <td>IPI00108279;IPI00808211</td>\n",
       "      <td>hypothetical protein LOC73533;hypothetical pro...</td>\n",
       "      <td>_APTRS(ph)VNGEIWYKAIW_</td>\n",
       "      <td>1700080G18Rik</td>\n",
       "      <td>Putative uncharacterized protein</td>\n",
       "      <td>1.435810e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8030462N17Rik</td>\n",
       "      <td>S</td>\n",
       "      <td>143</td>\n",
       "      <td>IPI00466384</td>\n",
       "      <td>Class I</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>GRKSRRSRSESET</td>\n",
       "      <td>3.41</td>\n",
       "      <td>Positive</td>\n",
       "      <td>2.73</td>\n",
       "      <td>...</td>\n",
       "      <td>256.52</td>\n",
       "      <td>Q8BH50-1;Q8BH50;Q0VAW6;Q19SD0;Q8BH50-2;Q19SC5</td>\n",
       "      <td>IPI00466384;IPI00874579;IPI00719980</td>\n",
       "      <td>143</td>\n",
       "      <td>IPI00466384</td>\n",
       "      <td>RIKEN cDNA 8030462N17, isoform CRA_a;Isoform 2...</td>\n",
       "      <td>_S(ph)RS(ph)ES(ph)ETSTMAAK_</td>\n",
       "      <td>8030462N17Rik;mCG_18277;Akd2;8030462N17Rik;Akd2</td>\n",
       "      <td>Uncharacterized protein C18orf25 homolog;Putat...</td>\n",
       "      <td>6.170000e-55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8030462N17Rik</td>\n",
       "      <td>S</td>\n",
       "      <td>67</td>\n",
       "      <td>IPI00466384</td>\n",
       "      <td>Class I</td>\n",
       "      <td>0.999990</td>\n",
       "      <td>ELRRDSSESQLAS</td>\n",
       "      <td>-1.09</td>\n",
       "      <td>Negative</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>...</td>\n",
       "      <td>257.50</td>\n",
       "      <td>Q8BH50-1;Q8BH50;Q0VAW6;Q19SD0;Q8BH50-2</td>\n",
       "      <td>IPI00466384;IPI00874579</td>\n",
       "      <td>67</td>\n",
       "      <td>IPI00466384</td>\n",
       "      <td>RIKEN cDNA 8030462N17, isoform CRA_a;Isoform 2...</td>\n",
       "      <td>_RDS(ph)S(ph)ESQIASTESDKPTTGR_</td>\n",
       "      <td>8030462N17Rik;mCG_18277;Akd2</td>\n",
       "      <td>Uncharacterized protein C18orf25 homolog;Putat...</td>\n",
       "      <td>5.960000e-26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene Name Phosphorylated Amino Acid  IPI 3.68 Position in Protein  \\\n",
       "0          39326                         T                           426   \n",
       "1  1600027N09Rik                         S                           191   \n",
       "2  1700080G18Rik                         S                           308   \n",
       "3  8030462N17Rik                         S                           143   \n",
       "4  8030462N17Rik                         S                            67   \n",
       "\n",
       "           IPI Site Localization Class  Localization Probability  \\\n",
       "0  IPI00224626                 Class I                  0.999318   \n",
       "1  IPI00720110                 Class I                  0.999998   \n",
       "2  IPI00108279                 Class I                  0.998163   \n",
       "3  IPI00466384                 Class I                  1.000000   \n",
       "4  IPI00466384                 Class I                  0.999990   \n",
       "\n",
       "  Sequence Window  Inhibitor Exps Insulin/Starved Median (log2)  \\\n",
       "0   QQNSSRTLEKNKK                                         -1.41   \n",
       "1   KVLTANSNPSSPS                                         -2.41   \n",
       "2   FKAPTRSVNGELW                                         -2.64   \n",
       "3   GRKSRRSRSESET                                          3.41   \n",
       "4   ELRRDSSESQLAS                                         -1.09   \n",
       "\n",
       "  LY Screen Insulin Regulated Direction  \\\n",
       "0                              Negative   \n",
       "1                              Negative   \n",
       "2                              Negative   \n",
       "3                              Positive   \n",
       "4                              Negative   \n",
       "\n",
       "   LY Exps Insulin/Starved Median (log2)  ...  Andromeda Search Score  \\\n",
       "0                                  -1.46  ...                  261.67   \n",
       "1                                  -2.74  ...                  195.38   \n",
       "2                                  -2.27  ...                   70.98   \n",
       "3                                   2.73  ...                  256.52   \n",
       "4                                  -1.00  ...                  257.50   \n",
       "\n",
       "                                      Uniprot ID  \\\n",
       "0             Q3UGE0;Q3UNN1;Q5DTS3;Q8C2A3;O55131   \n",
       "1                                  Q14AR7;Q2NL54   \n",
       "2                                         Q9CVQ8   \n",
       "3  Q8BH50-1;Q8BH50;Q0VAW6;Q19SD0;Q8BH50-2;Q19SC5   \n",
       "4         Q8BH50-1;Q8BH50;Q0VAW6;Q19SD0;Q8BH50-2   \n",
       "\n",
       "                      Proteins IPI IDs  Positions         Leading Proteins  \\\n",
       "0              IPI00224626;IPI00874440        426              IPI00224626   \n",
       "1                          IPI00720110        191              IPI00720110   \n",
       "2              IPI00108279;IPI00808211    308;382  IPI00108279;IPI00808211   \n",
       "3  IPI00466384;IPI00874579;IPI00719980        143              IPI00466384   \n",
       "4              IPI00466384;IPI00874579         67              IPI00466384   \n",
       "\n",
       "                                Protein Descriptions  \\\n",
       "0            cell division cycle 10 homolog;Septin-7   \n",
       "1                         RIKEN cDNA 1600027N09 gene   \n",
       "2  hypothetical protein LOC73533;hypothetical pro...   \n",
       "3  RIKEN cDNA 8030462N17, isoform CRA_a;Isoform 2...   \n",
       "4  RIKEN cDNA 8030462N17, isoform CRA_a;Isoform 2...   \n",
       "\n",
       "                Modified Sequence  \\\n",
       "0             _IIEQQNSSRT(ph)IEK_   \n",
       "1       _VITANS(ph)NPSS(ph)PSAAK_   \n",
       "2          _APTRS(ph)VNGEIWYKAIW_   \n",
       "3     _S(ph)RS(ph)ES(ph)ETSTMAAK_   \n",
       "4  _RDS(ph)S(ph)ESQIASTESDKPTTGR_   \n",
       "\n",
       "                                          Gene Names  \\\n",
       "0                        Sept7;mKIAA4020;Cdc10;Sept7   \n",
       "1  1600027N09Rik;mCG_6726;RP23-74K21.4;RP23-74K21...   \n",
       "2                                      1700080G18Rik   \n",
       "3    8030462N17Rik;mCG_18277;Akd2;8030462N17Rik;Akd2   \n",
       "4                       8030462N17Rik;mCG_18277;Akd2   \n",
       "\n",
       "                                       Protein Names           PEP  \n",
       "0  Putative uncharacterized protein;MKIAA4020 pro...  2.730000e-55  \n",
       "1  MCG6726, isoform CRA_f;Novel protein (Possible...  1.960000e-12  \n",
       "2                   Putative uncharacterized protein  1.435810e-03  \n",
       "3  Uncharacterized protein C18orf25 homolog;Putat...  6.170000e-55  \n",
       "4  Uncharacterized protein C18orf25 homolog;Putat...  5.960000e-26  \n",
       "\n",
       "[5 rows x 44 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "published_human_url = 'https://ars.els-cdn.com/content/image/1-s2.0-S1550413113001526-mmc2.xlsx'\n",
    "published_human_data = os.path.join(test_dir,'human_phosphosites.xlsx')\n",
    "\n",
    "if not os.path.exists(published_human_data):\n",
    "    print('Downloading human phosphoproteome data')\n",
    "    download_file(published_human_url, published_human_data)\n",
    "\n",
    "df = pd.read_excel(published_human_data, sheet_name='Inhibitor screen regulated')\n",
    "display(df.head())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pass\n",
      "True    1798\n",
      "Name: count, dtype: int64\n",
      "8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot ID</th>\n",
       "      <th>Phosphorylated Amino Acid</th>\n",
       "      <th>Sequence Window</th>\n",
       "      <th>IPI 3.68 Position in Protein</th>\n",
       "      <th>Inhibitor Exps Insulin/Starved Median (log2)</th>\n",
       "      <th>middle</th>\n",
       "      <th>pass</th>\n",
       "      <th>ID List</th>\n",
       "      <th>motif</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q3UGE0;Q3UNN1;Q5DTS3;Q8C2A3;O55131</td>\n",
       "      <td>T</td>\n",
       "      <td>QQNSSRTLEKNKK</td>\n",
       "      <td>426</td>\n",
       "      <td>-1.41</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>Q5DTS3</td>\n",
       "      <td>QNSSRTLEKN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q3UGE0;Q3UNN1;Q5DTS3;Q8C2A3;O55131</td>\n",
       "      <td>T</td>\n",
       "      <td>QQNSSRTLEKNKK</td>\n",
       "      <td>426</td>\n",
       "      <td>-1.41</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>Q3UNN1</td>\n",
       "      <td>QNSSRTLEKN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q3UGE0;Q3UNN1;Q5DTS3;Q8C2A3;O55131</td>\n",
       "      <td>T</td>\n",
       "      <td>QQNSSRTLEKNKK</td>\n",
       "      <td>426</td>\n",
       "      <td>-1.41</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>O55131</td>\n",
       "      <td>QNSSRTLEKN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q3UGE0;Q3UNN1;Q5DTS3;Q8C2A3;O55131</td>\n",
       "      <td>T</td>\n",
       "      <td>QQNSSRTLEKNKK</td>\n",
       "      <td>426</td>\n",
       "      <td>-1.41</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>Q8C2A3</td>\n",
       "      <td>QNSSRTLEKN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q3UGE0;Q3UNN1;Q5DTS3;Q8C2A3;O55131</td>\n",
       "      <td>T</td>\n",
       "      <td>QQNSSRTLEKNKK</td>\n",
       "      <td>426</td>\n",
       "      <td>-1.41</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>Q3UGE0</td>\n",
       "      <td>QNSSRTLEKN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Uniprot ID Phosphorylated Amino Acid  \\\n",
       "0  Q3UGE0;Q3UNN1;Q5DTS3;Q8C2A3;O55131                         T   \n",
       "1  Q3UGE0;Q3UNN1;Q5DTS3;Q8C2A3;O55131                         T   \n",
       "2  Q3UGE0;Q3UNN1;Q5DTS3;Q8C2A3;O55131                         T   \n",
       "3  Q3UGE0;Q3UNN1;Q5DTS3;Q8C2A3;O55131                         T   \n",
       "4  Q3UGE0;Q3UNN1;Q5DTS3;Q8C2A3;O55131                         T   \n",
       "\n",
       "  Sequence Window  IPI 3.68 Position in Protein  \\\n",
       "0   QQNSSRTLEKNKK                           426   \n",
       "1   QQNSSRTLEKNKK                           426   \n",
       "2   QQNSSRTLEKNKK                           426   \n",
       "3   QQNSSRTLEKNKK                           426   \n",
       "4   QQNSSRTLEKNKK                           426   \n",
       "\n",
       "   Inhibitor Exps Insulin/Starved Median (log2) middle  pass ID List  \\\n",
       "0                                         -1.41      T  True  Q5DTS3   \n",
       "1                                         -1.41      T  True  Q3UNN1   \n",
       "2                                         -1.41      T  True  O55131   \n",
       "3                                         -1.41      T  True  Q8C2A3   \n",
       "4                                         -1.41      T  True  Q3UGE0   \n",
       "\n",
       "        motif  \n",
       "0  QNSSRTLEKN  \n",
       "1  QNSSRTLEKN  \n",
       "2  QNSSRTLEKN  \n",
       "3  QNSSRTLEKN  \n",
       "4  QNSSRTLEKN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot_ID</th>\n",
       "      <th>Position</th>\n",
       "      <th>Motif</th>\n",
       "      <th>log2FC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q5DTS3</td>\n",
       "      <td>426</td>\n",
       "      <td>QNSSRTLEKN</td>\n",
       "      <td>-1.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q3UNN1</td>\n",
       "      <td>426</td>\n",
       "      <td>QNSSRTLEKN</td>\n",
       "      <td>-1.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O55131</td>\n",
       "      <td>426</td>\n",
       "      <td>QNSSRTLEKN</td>\n",
       "      <td>-1.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q8C2A3</td>\n",
       "      <td>426</td>\n",
       "      <td>QNSSRTLEKN</td>\n",
       "      <td>-1.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q3UGE0</td>\n",
       "      <td>426</td>\n",
       "      <td>QNSSRTLEKN</td>\n",
       "      <td>-1.41</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Uniprot_ID  Position       Motif  log2FC\n",
       "0     Q5DTS3       426  QNSSRTLEKN   -1.41\n",
       "1     Q3UNN1       426  QNSSRTLEKN   -1.41\n",
       "2     O55131       426  QNSSRTLEKN   -1.41\n",
       "3     Q8C2A3       426  QNSSRTLEKN   -1.41\n",
       "4     Q3UGE0       426  QNSSRTLEKN   -1.41"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of phosphosites: 5936\n"
     ]
    }
   ],
   "source": [
    "df = df[['Uniprot ID', 'Phosphorylated Amino Acid', 'Sequence Window', 'IPI 3.68 Position in Protein', 'Inhibitor Exps Insulin/Starved Median (log2)']].copy()\n",
    "df['middle'] = df['Sequence Window'].apply(lambda x: x[6])\n",
    "df['pass'] = df['middle'] == df['Phosphorylated Amino Acid']\n",
    "\n",
    "#how many did not pass\n",
    "print(df['pass'].value_counts())\n",
    "\n",
    "#how many Uniprot IDs are blank\n",
    "print(df['Uniprot ID'].isnull().sum())\n",
    "\n",
    "#remove the empty Uniprot IDs and the ones that did not pass\n",
    "df = df[(df['Uniprot ID'].notnull()) & df['pass']].copy()\n",
    "\n",
    "df['ID List'] = df['Uniprot ID'].apply(lambda x: set([y.split('-')[0] for y in x.split(';')]))\n",
    "data_exploded_df = df.explode('ID List').reset_index(drop=True)\n",
    "data_exploded_df['motif'] = data_exploded_df['Sequence Window'].apply(lambda x: x[1:11])\n",
    "\n",
    "display(data_exploded_df.head())\n",
    "\n",
    "df = data_exploded_df[['ID List', 'IPI 3.68 Position in Protein', 'motif', 'Inhibitor Exps Insulin/Starved Median (log2)']].copy()\n",
    "#rename columns to Uniprot_ID, Position, Motif, and log2FC\n",
    "df.columns = ['Uniprot_ID', 'Position', 'Motif', 'log2FC']\n",
    "\n",
    "#remove duplicate rows\n",
    "df = df.drop_duplicates()\n",
    "\n",
    "display(df.head())\n",
    "\n",
    "\n",
    "human_experiment_file = os.path.join(test_dir, 'human_experiment.csv')\n",
    "print(f'Number of phosphosites: {len(df)}')\n",
    "df.to_csv(human_experiment_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Duplicated IDs in ST data: 2053    Q9Z2H5_441\n",
      "3924    Q8BTY2_238\n",
      "Name: __ENTRY_ID, dtype: object\n",
      "Elapsed time for percentiles and matches: 17.48 seconds\n"
     ]
    }
   ],
   "source": [
    "column_names_dict={'id':'Uniprot_ID', 'site':'Position', 'peptide':'Motif', 'log2fc':'log2FC', 'dependent': None}\n",
    "\n",
    "df = pd.read_csv(human_experiment_file)\n",
    "\n",
    "session = Session(0, 'human', df, column_names_dict, scoring, background, ortholog_manager, id_type='UniProtKB', ambiguous=True, debug=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
